src/launch/linux/ioctl.rs

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::impl_const_id;
+use crate::launch::types::*;
+use iocuddle::*;
+
+use std::convert::TryInto;
+use std::io::{Error, Result};
+use std::marker::PhantomData;
+use std::os::raw::{c_int, c_uint, c_ulong};
+use std::os::unix::io::AsRawFd;
+
+// These enum ordinal values are defined in the Linux kernel
+// source code: include/uapi/linux/kvm.h
+impl_const_id! {
+    pub Id => u32;
+    Init = 0,
+    LaunchStart<'_> = 2,
+    LaunchUpdateData<'_> = 3,
+    LaunchSecret<'_> = 5,
+    LaunchMeasure<'_> = 6,
+    LaunchFinish = 7,
+}
+
+const ENC_OP: u64 = 0xc008aeba;
+const REG_REGION: u64 = 0x8010aebb;
+
+// Note: the iocuddle::Ioctl::classic constructor has been used here because
+// KVM_MEMORY_ENCRYPT_OP ioctl was defined like this:
+//
+// _IOWR(KVMIO, 0xba, unsigned long)
+//
+// Instead of something like this:
+//
+// _IOWR(KVMIO, 0xba, struct kvm_sev_cmd)
+//
+// which would require extra work to wrap around the design decision for
+// that ioctl.
+
+/// Initialize the SEV platform context.
+pub const INIT: Ioctl<WriteRead, &Command<Init>> = unsafe { Ioctl::classic(ENC_OP) };
+/// Create encrypted guest context.
+pub const LAUNCH_START: Ioctl<WriteRead, &Command<LaunchStart>> = unsafe { Ioctl::classic(ENC_OP) };
+/// Encrypt guest data with its VEK.
+pub const LAUNCH_UPDATE_DATA: Ioctl<WriteRead, &Command<LaunchUpdateData>> =
+    unsafe { Ioctl::classic(ENC_OP) };
+/// Inject a secret into the guest.
+pub const LAUNCH_SECRET: Ioctl<WriteRead, &Command<LaunchSecret>> =
+    unsafe { Ioctl::classic(ENC_OP) };
+/// Get the guest's measurement.
+pub const LAUNCH_MEASUREMENT: Ioctl<WriteRead, &Command<LaunchMeasure>> =
+    unsafe { Ioctl::classic(ENC_OP) };
+/// Complete the SEV launch flow and transition the guest into
+/// the ready state.
+pub const LAUNCH_FINISH: Ioctl<WriteRead, &Command<LaunchFinish>> =
+    unsafe { Ioctl::classic(ENC_OP) };
+
+#[repr(C)]
+pub struct EncryptedRegion<'a> {
+    addr: u64,
+    len: u32,
+    _phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> EncryptedRegion<'a> {
+    pub fn new(region: &'a [u8]) -> Self {
+        Self {
+            addr: region.as_ptr() as _,
+            len: region.len() as _,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn ioctl(self, fd: &impl AsRawFd) -> Result<c_uint> {
```

Comment:
Don't you want to add this to `new()`? The function name `ioctl` doesn't say much. What does it do?

Reply:
Actually, I changed my mind. The `EncryptedRegion` is a stand-in rough replacement for an `iocuddle` wrapper, but since one can't be created, this is the interface we get. I'll leave construction and ioctl as 2 separate steps.

Reply:
Should it follow the builder pattern then? Can I call `ioctl()` more than once? Is the `EncryptedRegion` encrypted without the `ioctl`?

Reply:
I see where you're going with this line of questioning. I'm wondering if the `ioctl` method is better expressed as a free function so that usage would look more like:

```rust
let region = EncryptedRegion { .. };
EncryptedRegion::ioctl(&region).unwrap();
```

The `EncryptedRegion` is just an internal interface that groups the parameters expected by the Linux kernel in a struct defined by the Linux kernel.

Reply:
What do you think of the public interface being a:
```rust
pub fn encrypt_region(region: &'a [u8], kvm_fd: &impl AsRawFd) -> Result<c_uint> {
    EncryptedRegion::new(region).ioctl(kvm_fd)
}
``` 
and the rest is private?

Reply:
I like it! I'll queue this change